warehouse/accounts/utils.py

```diff
@@ -0,0 +1,181 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import base64
+import re
+
+import attr
+
+from warehouse.email import send_token_compromised_email_leak
+from warehouse.macaroons.caveats import InvalidMacaroon
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.metrics import IMetricsService
+
+
+class ExtractionFailed(Exception):
+    pass
+
+
+class TokenLeakMatcher:
+    """
+    A TokenLeakMatcher is linked to a specific regex pattern. When provided
+    a string that matches this pattern, the matcher can extract a token-like string
+    from it.
+    """
+
+    name: str
+    pattern: re.Pattern
+
+    def extract(self, text):
+        raise NotImplementedError
+
+
+class PlainTextTokenLeakMatcher(TokenLeakMatcher):
+    name = "token"
+    # Macaroons are urlsafe_b64 encodeded so non-alphanumeric chars are - and _
+    # https://github.com/ecordell/pymacaroons/blob/06b55110eda2fb192c130dee0bcedf8b124d1056/pymacaroons/serializers/binary_serializer.py#L32
+    pattern = re.compile(r"pypi-[A-Za-z0-9-_=]+")
+
+    def extract(self, text):
+        """
+        From a string containing everything that was matched, extract the token
+        to check
+        """
+        return text
+
+
+class Base64BasicAuthTokenLeakMatcher(TokenLeakMatcher):
+    name = "base64-basic-auth"
+    # This is what we would expect to find if a basic auth value was leaked
+    # The following string was obtained by:
+    #     base64.b64encode(b"__token__:pypi-").decode("utf-8")
+    # Basic auth is standard base64, so non-alphanumeric chars are + and /
+    pattern = re.compile(r"X190b2tlbl9fOnB5cGkt[A-Za-z0-9+/=]+")
+
+    def extract(self, text):
+        try:
+            _, token = (
+                base64.b64decode(text.encode("utf-8")).decode("utf-8").split(":", 1)
+            )
+            return token
+        except Exception as exc:
+            raise ExtractionFailed from exc
+
+
+TOKEN_LEAK_MATCHERS = {
+    matcher.name: matcher
+    for matcher in [PlainTextTokenLeakMatcher(), Base64BasicAuthTokenLeakMatcher()]
+}
+
+
+class InvalidTokenLeakRequest(Exception):
+    def __init__(self, message, reason):
+        self.reason = reason
+        super().__init__(message)
+
+
+@attr.s(auto_attribs=True, frozen=True, slots=True)
```

Comment:
Currently `attrs` is not a top-level dependency of Warehouse, it's only available as a sub-dependency of `automat`. If we truly need to use `attrs` here, it should be added to https://github.com/pypa/warehouse/blob/master/requirements/main.in and those dependencies should be re-compiled.

However I think I'd prefer to just use a regular class here.

Reply:
Hm, I’m 100% sure that when this code was written, I copy pasted this from elsewhere in the codebase, so either it was removed since then or something happened that I don’t understand, but I’m usually not in the habbit of adding unneeded deps. I’ll reove it.

Reply:
I'll remove it, but you may want to fix this in another ticket/PR:

https://github.com/pypa/warehouse/blob/b344984ee83126eb6fc541406d1f34719fad8b6b/warehouse/email/services.py#L36

Reply:
Ah, good catch!